In [1]:
import pandas as pd
import os
from pathlib import Path

In [2]:
# settings to display all columns and rows
pd.set_option("display.max_columns", None)

In [3]:
# Convert all .sas7bdat files to .xlsx files

In [4]:
def convert_sas7bdat_to_excel(inputpath):
    df = pd.read_sas(inputpath)
    df = df.copy()
    outputpath = Path(inputpath.parent, inputpath.stem + ".xlsx")
    if os.path.exists(outputpath):
        print(".xlsx already exists, skipping this .sas7bdat file")
    else:
        for i in df:
            if df[i].dtype == 'O':
                df[i] = df[i].str.decode('UTF-8')
        print("df.shape:", df.shape)
        if df.shape[0] <= 1000000:
            df.to_excel(outputpath, index=False)
            print("df saved to excel")
        else:
            print("df has > 1 mil rows, cannot be saved to excel")

In [5]:
inputdir = Path(r"C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data")

# Part 1

In [21]:
for inputfilename in os.listdir(inputdir):
    if "sas7bdat" in inputfilename:
        inputfilepath = Path(inputdir / inputfilename)
        print(inputfilepath)
        convert_sas7bdat_to_excel(inputfilepath)

C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\anncomp.sas7bdat


C:\Users\jasonjia\Anaconda3\lib\site-packages\pandas\io\sas\sas7bdat.py:800: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`
  rslt[name] = self._byte_chunk[jb, :].view(dtype=self.byte_order + "d")


.xlsx already exists, skipping this .sas7bdat file
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\codirfin.sas7bdat
df.shape: (26187, 71)
df saved to excel
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\colev.sas7bdat
df.shape: (3915, 18)
df saved to excel
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\coperol.sas7bdat
df.shape: (63403, 14)
df saved to excel
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\deferredcomp.sas7bdat
df.shape: (60331, 46)
df saved to excel
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\directorcomp.sas7bdat
df.shape: (142230, 28)
df saved to excel
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\exnames.sas7bdat
df.shape: (3906, 5)
df saved to excel
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\ex_black.sas7bdat
df.shape: (15, 6)
df saved to

C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\anncomp.sas7bdat

C:\Users\jasonjia\Anaconda3\lib\site-packages\pandas\io\sas\sas7bdat.py:800: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`
  rslt[name] = self._byte_chunk[jb, :].view(dtype=self.byte_order + "d")

.xlsx already exists, skipping this .sas7bdat file

C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\codirfin.sas7bdat
df.shape: (26187, 71)
df saved to excel
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\colev.sas7bdat
df.shape: (3915, 18)
df saved to excel
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\coperol.sas7bdat
df.shape: (63403, 14)
df saved to excel
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\deferredcomp.sas7bdat
df.shape: (60331, 46)
df saved to excel
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\directorcomp.sas7bdat
df.shape: (142230, 28)
df saved to excel
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\exnames.sas7bdat
df.shape: (3906, 5)
df saved to excel
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\ex_black.sas7bdat
df.shape: (15, 6)
df saved to excel
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\ex_header.sas7bdat
df.shape: (63459, 38)
df saved to excel
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\ltawdtab.sas7bdat
df.shape: (6606, 48)
df saved to excel
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\outstandingawards.sas7bdat
df.shape: (569005, 50)
df saved to excel
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\pension.sas7bdat
df.shape: (65362, 44)
df saved to excel
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\person.sas7bdat
df.shape: (56289, 8)
df saved to excel
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\stgrttab.sas7bdat
df.shape: (27198, 49)
df saved to excel


# Part 2

In [6]:
isinpath = Path(r"C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\original_isins\isins_original.xlsx")
outputdir = Path(r"C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\gvkey_found_execucomp_check")
resultpath = Path(r"C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\gvkey_found_execucomp_check\results.xlsx")

In [7]:
# Open up isins.xlsx
df_isin = pd.read_excel(isinpath)
isin_gvkey = df_isin['GVKEY']
print("isin_gvkey shape:", isin_gvkey.shape)
# There should be no NAs
isin_gvkey = isin_gvkey.dropna()
print("isin_gvkey shape after dropping na:", isin_gvkey.shape)

isin_gvkey shape: (4140,)
isin_gvkey shape after dropping na: (4140,)


In [8]:
# get unique gvkeys in isins.xlsx and compute %
isin_gvkey_unique = isin_gvkey.unique()
print("isin_gvkey_unique shape:", isin_gvkey_unique.shape)
isin_gvkey_unique

isin_gvkey_unique shape: (3424,)


array([  2270.,   3221.,   3497., ..., 102422., 210057., 211655.])

In [9]:
results = pd.DataFrame(columns=['name_of_execucomp_file', '% of gvkeys found out of 4140 rows', '% of gvkeys found out of 3424 unique gvkeys (from 4140 rows)'])

In [20]:
for inputfilename in os.listdir(inputdir):
    if "xlsx" in inputfilename:
        # Get inputfilepath
        inputfilepath = Path(inputdir / inputfilename)
        print(inputfilepath)
        outputfilename = inputfilepath.stem + "_gvkey_found.xlsx"
        outputfilepath = Path(outputdir / outputfilename)
        
        # Import an execucomp excel file
        df_execucomp = pd.read_excel(inputfilepath)
        
        # Check file only if it has a GVKEY column
        if any('GVKEY' in column_name for column_name in df_execucomp.columns):
            # Get the gvkeys of the execucomp excel file
            input_gvkey = df_execucomp['GVKEY']
            print("input_gvkey shape:", input_gvkey.shape)
            input_gvkey = input_gvkey.dropna() 
            print("input_gvkey shape after dropping na:", input_gvkey.shape) # There should be no NAs

            # Get unique gvkeys in execucomp excel file
            input_gvkey_unique = input_gvkey.unique()
            print("input_gvkey_unique shape:", input_gvkey_unique.shape)

            # Compare to check if each (non-unique) gvkey in isins.xlsx is in the execucomp gvkeys
            gvkey_found = isin_gvkey.isin(input_gvkey_unique)
            df_isin['gvkey_found'] = gvkey_found
            pct_gvkey_found = round(gvkey_found.mean() * 100, 1)
            print("% of gvkeys found out of 4140 rows:", pct_gvkey_found)

            # Compare to check if each unique gvkey in isins.xlsx is in the execucomp gvkeys
            gvkey_found_unique = pd.DataFrame(isin_gvkey_unique).isin(input_gvkey_unique)
            pct_gvkey_found_unique = round(gvkey_found_unique.mean()[0] * 100, 1)
            print("% of gvkeys found out of 3424 unique gvkeys (from 4140 rows):", pct_gvkey_found_unique)

            # Create df_isin_unique: isins.xlsx with duplicate gvkeys removed
            df_isin_unique = df_isin.drop_duplicates(subset=['GVKEY'])
            df_isin_unique.reset_index(drop=True, inplace=True)

            # Add gvkey_found column to df_isin_unique
            df_isin_unique['gvkey_found'] = gvkey_found_unique

            # Save matches into an excel file
            with pd.ExcelWriter(outputfilepath) as writer:  
                df_isin_unique.to_excel(writer, sheet_name='gvkey_found_unique', index=False)
                df_isin.to_excel(writer, sheet_name='gvkey_found_original', index=False)

            # Append % gvkeys found statistics to result_df
            result_row = {'name_of_execucomp_file': str(inputfilepath),
                          '% of gvkeys found out of 4140 rows': pct_gvkey_found,
                          '% of gvkeys found out of 3424 unique gvkeys (from 4140 rows)': pct_gvkey_found_unique}
            results = results.append(result_row, ignore_index = True)
        else:
            # Append 'No gvkey column' to result_df
            result_row = {'name_of_execucomp_file': str(inputfilepath),
                          '% of gvkeys found out of 4140 rows': 'No gvkey column',
                          '% of gvkeys found out of 3424 unique gvkeys (from 4140 rows)': 'No gvkey column'}
            results = results.append(result_row, ignore_index = True)

C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\anncomp.xlsx
input_gvkey shape: (116190,)
input_gvkey shape after dropping na: (116190,)
input_gvkey_unique shape: (1246,)
% of gvkeys found out of 4140 rows: 32.8
% of gvkeys found out of 3424 unique gvkeys (from 4140 rows): 36.4


C:\Users\jasonjia\AppData\Local\Temp/ipykernel_21188/3891543425.py:40: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_isin_unique['gvkey_found'] = gvkey_found_unique


C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\codirfin.xlsx
input_gvkey shape: (26187,)
input_gvkey shape after dropping na: (26187,)
input_gvkey_unique shape: (1247,)
% of gvkeys found out of 4140 rows: 32.8
% of gvkeys found out of 3424 unique gvkeys (from 4140 rows): 36.4
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\colev.xlsx
input_gvkey shape: (3915,)
input_gvkey shape after dropping na: (3915,)
input_gvkey_unique shape: (3915,)
% of gvkeys found out of 4140 rows: 32.8
% of gvkeys found out of 3424 unique gvkeys (from 4140 rows): 36.4
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\coperol.xlsx
input_gvkey shape: (63403,)
input_gvkey shape after dropping na: (63403,)
input_gvkey_unique shape: (3906,)
% of gvkeys found out of 4140 rows: 32.8
% of gvkeys found out of 3424 unique gvkeys (from 4140 rows): 36.4
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\

TypeError: to_excel() got an unexpected keyword argument 'ignore_index'

C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\anncomp.xlsx
input_gvkey shape: (116190,)
input_gvkey shape after dropping na: (116190,)
input_gvkey_unique shape: (1246,)
% of gvkeys found out of 4140 rows: 32.8
% of gvkeys found out of 3424 unique gvkeys (from 4140 rows): 36.4
C:\Users\jasonjia\AppData\Local\Temp/ipykernel_21188/3891543425.py:40: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_isin_unique['gvkey_found'] = gvkey_found_unique
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\codirfin.xlsx
input_gvkey shape: (26187,)
input_gvkey shape after dropping na: (26187,)
input_gvkey_unique shape: (1247,)
% of gvkeys found out of 4140 rows: 32.8
% of gvkeys found out of 3424 unique gvkeys (from 4140 rows): 36.4
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\colev.xlsx
input_gvkey shape: (3915,)
input_gvkey shape after dropping na: (3915,)
input_gvkey_unique shape: (3915,)
% of gvkeys found out of 4140 rows: 32.8
% of gvkeys found out of 3424 unique gvkeys (from 4140 rows): 36.4
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\coperol.xlsx
input_gvkey shape: (63403,)
input_gvkey shape after dropping na: (63403,)
input_gvkey_unique shape: (3906,)
% of gvkeys found out of 4140 rows: 32.8
% of gvkeys found out of 3424 unique gvkeys (from 4140 rows): 36.4
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\deferredcomp.xlsx
input_gvkey shape: (60331,)
input_gvkey shape after dropping na: (60331,)
input_gvkey_unique shape: (952,)
% of gvkeys found out of 4140 rows: 25.2
% of gvkeys found out of 3424 unique gvkeys (from 4140 rows): 27.8
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\directorcomp.xlsx
input_gvkey shape: (142230,)
input_gvkey shape after dropping na: (142230,)
input_gvkey_unique shape: (1215,)
% of gvkeys found out of 4140 rows: 32.0
% of gvkeys found out of 3424 unique gvkeys (from 4140 rows): 35.5
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\exnames.xlsx
input_gvkey shape: (3906,)
input_gvkey shape after dropping na: (3906,)
input_gvkey_unique shape: (3906,)
% of gvkeys found out of 4140 rows: 32.8
% of gvkeys found out of 3424 unique gvkeys (from 4140 rows): 36.4
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\ex_black.xlsx
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\ex_header.xlsx
input_gvkey shape: (63459,)
input_gvkey shape after dropping na: (63403,)
input_gvkey_unique shape: (3906,)
% of gvkeys found out of 4140 rows: 32.8
% of gvkeys found out of 3424 unique gvkeys (from 4140 rows): 36.4
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\ltawdtab.xlsx
input_gvkey shape: (6606,)
input_gvkey shape after dropping na: (6606,)
input_gvkey_unique shape: (332,)
% of gvkeys found out of 4140 rows: 8.8
% of gvkeys found out of 3424 unique gvkeys (from 4140 rows): 9.7
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\outstandingawards.xlsx
input_gvkey shape: (569005,)
input_gvkey shape after dropping na: (569005,)
input_gvkey_unique shape: (1211,)
% of gvkeys found out of 4140 rows: 31.9
% of gvkeys found out of 3424 unique gvkeys (from 4140 rows): 35.4
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\pension.xlsx
input_gvkey shape: (65362,)
input_gvkey shape after dropping na: (65362,)
input_gvkey_unique shape: (657,)
% of gvkeys found out of 4140 rows: 17.5
% of gvkeys found out of 3424 unique gvkeys (from 4140 rows): 19.2
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\person.xlsx
C:\Users\jasonjia\Dropbox\Projects\boardex_and_execucomp\output\execucomp_data\stgrttab.xlsx
input_gvkey shape: (27198,)
input_gvkey shape after dropping na: (27198,)
input_gvkey_unique shape: (830,)
% of gvkeys found out of 4140 rows: 22.4
% of gvkeys found out of 3424 unique gvkeys (from 4140 rows): 24.2


In [23]:
results.to_excel(resultpath, index=False)
print("Done!")

Done!
